File: android-core/src/androidTest/java/com/mparticle/internal/KitFrameworkWrapperTest.java
package com.mparticle.internal;

import android.content.Context;
import android.util.MutableBoolean;

import com.mparticle.AccessUtils;
import com.mparticle.MParticle;
import com.mparticle.identity.IdentityApiRequest;
import com.mparticle.identity.MParticleUser;
import com.mparticle.testutils.BaseCleanStartedEachTest;
import com.mparticle.testutils.MPLatch;

import org.junit.Test;
import org.mockito.Mockito;

import java.util.Random;
import java.util.concurrent.CountDownLatch;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertTrue;

public class KitFrameworkWrapperTest extends BaseCleanStartedEachTest {

    private void setKitManager(KitFrameworkWrapper kitManager) {
        AccessUtils.setKitManager(kitManager);
        com.mparticle.identity.AccessUtils.setKitManager(kitManager);
    }

    @Test
    public void testIdentify() throws InterruptedException {
        final Long mpid = new Random().nextLong();
        final CountDownLatch latch = new MPLatch(1);
        final MutableBoolean called = new MutableBoolean(false);
        setKitManager(new StubKitManager(mContext) {
            @Override
            public void onIdentifyCompleted(MParticleUser user, IdentityApiRequest request) {
                if (user.getId() == mStartingMpid) {
                    return;
                }
                assertEquals(mpid.longValue(), user.getId());
                called.value = true;
                latch.countDown();
            }
        });
        mServer.setupHappyIdentify(mpid);
        MParticle.getInstance().Identity().identify(IdentityApiRequest.withEmptyUser().build());
        latch.await();
        assertTrue(called.value);
    }

    @Test
    public void testLogin() throws InterruptedException {
        final Long mpid = new Random().nextLong();
        final CountDownLatch latch = new MPLatch(1);
        final MutableBoolean called = new MutableBoolean(false);
        setKitManager(new StubKitManager(mContext) {
            @Override
            public void onLoginCompleted(MParticleUser user, IdentityApiRequest request) {
                if (user.getId() == mStartingMpid) {
                    return;
                }
                assertEquals(mpid.longValue(), user.getId());
                called.value = true;
                latch.countDown();
            }
        });
        mServer.setupHappyLogin(mpid);
        MParticle.getInstance().Identity().login(IdentityApiRequest.withEmptyUser().build());
        latch.await();
        assertTrue(called.value);
    }

    @Test
    public void testLogout() throws InterruptedException {
        final Long mpid = new Random().nextLong();
        final CountDownLatch latch = new MPLatch(1);
        final MutableBoolean called = new MutableBoolean(false);
        setKitManager(new StubKitManager(mContext) {
            @Override
            public void onLogoutCompleted(MParticleUser user, IdentityApiRequest request) {
                if (user.getId() == mStartingMpid) {
                    return;
                }
                assertEquals(mpid.longValue(), user.getId());
                called.value = true;
                latch.countDown();
            }
        });
        mServer.setupHappyLogout(mpid);
        MParticle.getInstance().Identity().logout(IdentityApiRequest.withEmptyUser().build());
        latch.await();
        assertTrue(called.value);
    }

    @Test
    public void testModify() throws InterruptedException {
        final CountDownLatch latch = new MPLatch(1);
        final MutableBoolean called = new MutableBoolean(false);
        setKitManager(new StubKitManager(mContext) {
            @Override
            public void onModifyCompleted(MParticleUser user, IdentityApiRequest request) {
                assertEquals(mStartingMpid.longValue(), user.getId());
                called.value = true;
                latch.countDown();
            }
        });
        MParticle.getInstance().Identity().modify(IdentityApiRequest.withUser(MParticle.getInstance().Identity().getCurrentUser()).build());
        latch.await();
        assertTrue(called.value);
    }

    @Test
    public void testModifyUserChanged() throws InterruptedException {
        final CountDownLatch latch = new MPLatch(1);
        final MutableBoolean called = new MutableBoolean(false);
        setKitManager(new StubKitManager(mContext) {
            @Override
            public void onModifyCompleted(MParticleUser user, IdentityApiRequest request) {
                assertEquals(mStartingMpid.longValue(), user.getId());
                called.value = true;
                latch.countDown();
            }
        });
        MParticle.getInstance().Identity().modify(IdentityApiRequest.withEmptyUser().build());
        MParticle.getInstance().getConfigManager().setMpid(0);
        latch.await();
        final CountDownLatch latch2 = new MPLatch(1);
        final Long mpid2 = new Random().nextLong();
        MParticle.getInstance().getConfigManager().setMpid(mpid2);
        assertTrue(called.value);
        called.value = false;
        setKitManager(new StubKitManager(mContext) {
            @Override
            public void onModifyCompleted(MParticleUser user, IdentityApiRequest request) {
                assertEquals(mpid2.longValue(), user.getId());
                called.value = true;
                latch2.countDown();
            }
        });
        MParticle.getInstance().Identity().modify(IdentityApiRequest.withUser(MParticle.getInstance().Identity().getCurrentUser()).build());
        MParticle.getInstance().getConfigManager().setMpid(new Random().nextLong());
        latch2.await();
        assertTrue(called.value);
    }

    static class StubKitManager extends KitFrameworkWrapper {

        public StubKitManager(Context context) {
            super(context, null, null, null, null, true);
            //using a Mockito instance for KitManager means that whenever a call is made to it, it will crash
            //this way we can override the methods we expect to be called, and if an unexpected one gets called, we will know about it
            setKitManager(Mockito.mock(KitManager.class));
        }

        @Override
        public void loadKitLibrary() {

        }
    }
}

File: android-core/src/androidTest/java/com/mparticle/internal/KitFrameworkWrapperTest.kt
package com.mparticle.internal

import android.content.Context
import android.util.MutableBoolean
import com.mparticle.AccessUtils
import com.mparticle.MParticle
import com.mparticle.identity.IdentityApiRequest
import com.mparticle.identity.MParticleUser
import com.mparticle.messages.IdentityResponseMessage
import com.mparticle.testing.BaseStartedTest
import com.mparticle.testing.FailureLatch
import com.mparticle.testing.Mutable
import com.mparticle.testing.context
import com.mparticle.testing.mockserver.EndpointType
import com.mparticle.testing.mockserver.Server
import com.mparticle.testing.mockserver.SuccessResponse
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.concurrent.CountDownLatch
import kotlin.random.Random

class KitFrameworkWrapperTest : BaseStartedTest() {
    private fun setKitManager(kitManager: KitFrameworkWrapper) {
        AccessUtils.setKitManager(kitManager)
        com.mparticle.identity.AccessUtils.setKitManager(kitManager)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testIdentify() {
        val mpid: Long = Random.Default.nextLong()
        val latch: CountDownLatch = FailureLatch()
        val called = Mutable(false)
        setKitManager(object : StubKitManager(context) {
            override fun onIdentifyCompleted(user: MParticleUser, request: IdentityApiRequest) {
                if (user.id == mStartingMpid) {
                    return
                }
                assertEquals(mpid, user.id)
                called.value = true
                latch.countDown()
            }
        })
        Server.endpoint(EndpointType.Identity_Identify)
            .addResponseLogic({ true }) {
                SuccessResponse { responseObject = IdentityResponseMessage(mpid = mpid) }
            }
        MParticle.getInstance()!!
            .Identity().identify(IdentityApiRequest.withEmptyUser().build())
        latch.await()
        Assert.assertTrue(called.value)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testLogin() {
        val mpid: Long = Random.Default.nextLong()
        val latch: CountDownLatch = FailureLatch()
        val called = Mutable(false)
        setKitManager(object : StubKitManager(context) {
            override fun onLoginCompleted(user: MParticleUser, request: IdentityApiRequest) {
                if (user.id == mStartingMpid) {
                    return
                }
                assertEquals(mpid, user.id)
                called.value = true
                latch.countDown()
            }
        })
        Server.endpoint(EndpointType.Identity_Login)
            .addResponseLogic({ true }) {
                SuccessResponse { responseObject = IdentityResponseMessage(mpid = mpid) }
            }
        MParticle.getInstance()!!
            .Identity().login(IdentityApiRequest.withEmptyUser().build())
        latch.await()
        Assert.assertTrue(called.value)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testLogout() {
        val mpid: Long = Random.Default.nextLong()
        val latch: CountDownLatch = FailureLatch()
        val called = Mutable(false)
        setKitManager(object : StubKitManager(context) {
            override fun onLogoutCompleted(user: MParticleUser, request: IdentityApiRequest) {
                if (user.id == mStartingMpid) {
                    return
                }
                assertEquals(mpid, user.id)
                called.value = true
                latch.countDown()
            }
        })
        Server.endpoint(EndpointType.Identity_Logout)
            .addResponseLogic({ true }) {
                SuccessResponse { responseObject = IdentityResponseMessage(mpid = mpid) }
            }
        MParticle.getInstance()!!
            .Identity().logout(IdentityApiRequest.withEmptyUser().build())
        latch.await()
        Assert.assertTrue(called.value)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testModify() {
        val latch: CountDownLatch = FailureLatch()
        val called = Mutable(false)
        setKitManager(object : StubKitManager(context) {
            override fun onModifyCompleted(user: MParticleUser, request: IdentityApiRequest) {
                assertEquals(mStartingMpid, user.id)
                called.value = true
                latch.countDown()
            }
        })
        MParticle.getInstance()!!.Identity().modify(
            IdentityApiRequest.withUser(
                MParticle.getInstance()!!.Identity().currentUser
            ).build()
        )
        latch.await()
        Assert.assertTrue(called.value)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testModifyUserChanged() {
        val latch: CountDownLatch = FailureLatch()
        val called = MutableBoolean(false)
        setKitManager(object : StubKitManager(context) {
            override fun onModifyCompleted(user: MParticleUser, request: IdentityApiRequest) {
                assertEquals(mStartingMpid, user.id)
                called.value = true
                latch.countDown()
            }
        })
        MParticle.getInstance()!!
            .Identity().modify(IdentityApiRequest.withEmptyUser().build())
        MParticle.getInstance()!!.Internal().configManager.setMpid(0, Random.Default.nextBoolean())
        latch.await()
        val latch2: CountDownLatch = FailureLatch()
        val mpid2: Long = Random.Default.nextLong()
        MParticle.getInstance()!!.Internal().configManager.setMpid(mpid2, Random.Default.nextBoolean())
        Assert.assertTrue(called.value)
        called.value = false
        setKitManager(object : StubKitManager(context) {
            override fun onModifyCompleted(user: MParticleUser, request: IdentityApiRequest) {
                assertEquals(mpid2, user.id)
                called.value = true
                latch2.countDown()
            }
        })
        MParticle.getInstance()!!.Identity().modify(
            IdentityApiRequest.withUser(
                MParticle.getInstance()!!.Identity().currentUser
            ).build()
        )
        MParticle.getInstance()!!
            .Internal().configManager.setMpid(Random.Default.nextLong(), Random.Default.nextBoolean())
        latch2.await()
        Assert.assertTrue(called.value)
    }

    internal open class StubKitManager(context: Context?) :
        KitFrameworkWrapper(context, null, null, null, true, null) {
        override fun loadKitLibrary() {}

        init {
            setKitManager(null)
        }
    }
}
